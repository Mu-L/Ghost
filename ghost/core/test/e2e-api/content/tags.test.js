const assert = require('assert/strict');
const {agentProvider, fixtureManager, matchers, dbUtils} = require('../../utils/e2e-framework');
const {anyContentVersion, anyEtag, anyObjectId} = matchers;

const matchTag = {
    id: anyObjectId
};

describe('Tags Content API', function () {
    let agent;

    before(async function () {
        agent = await agentProvider.getContentAPIAgent();
        await fixtureManager.init('users', 'user:inactive', 'posts', 'tags:extra', 'api_keys');
        agent.authenticate();
    });

    it('Can request tags', async function () {
        await agent
            .get('tags/')
            .expectStatus(200)
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            })
            .matchBodySnapshot({
                tags: new Array(4).fill(matchTag)
            })
            .expect((res) => {
                const {tags} = res.body;

                // Default order 'name asc' check
                // the ordering difference is described in https://github.com/TryGhost/Ghost/issues/6104
                if (dbUtils.isMySQL()) {
                    assert.equal(tags[0].name, 'bacon');
                    assert.equal(tags[3].name, 'kitchen sink');
                } else {
                    assert.equal(tags[0].name, 'Getting Started');
                    assert.equal(tags[3].name, 'kitchen sink');
                }
            });
    });

    it('Can request tags with limit=all', async function () {
        await agent
            .get('tags/?limit=all')
            .expectStatus(200)
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            })
            .matchBodySnapshot({
                tags: new Array(4).fill(matchTag)
            });
    });

    it('Can limit tags to receive', async function () {
        await agent
            .get('tags/?limit=3')
            .expectStatus(200)
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            })
            .matchBodySnapshot({
                tags: new Array(3).fill(matchTag)
            });
    });

    it('Can include post count', async function () {
        await agent
            .get('tags/?include=count.posts')
            .expectStatus(200)
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            })
            .matchBodySnapshot({
                tags: new Array(4).fill(matchTag)
            })
            .expect((res) => {
                const {tags} = res.body;

                // Each tag should have the correct count
                const getTagByName = name => tags.find(t => t.name === name);

                assert.equal(getTagByName('Getting Started').count.posts, 7);
                assert.equal(getTagByName('kitchen sink').count.posts, 2);
                assert.equal(getTagByName('bacon').count.posts, 2);
                assert.equal(getTagByName('chorizo').count.posts, 1);
            });
    });

    it('Can use multiple fields and have valid url fields', async function () {
        await agent
            .get('tags/?fields=url,name')
            .expectStatus(200)
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            })
            .matchBodySnapshot();
    });

    it('Can use single url field and have valid url fields', async function () {
        await agent
            .get('tags/?fields=url')
            .expectStatus(200)
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            })
            .matchBodySnapshot();
    });
});
