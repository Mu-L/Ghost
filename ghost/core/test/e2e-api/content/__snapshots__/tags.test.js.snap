// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags Content API Can include post count 1: [body] 1`] = `
Object {
  "meta": Object {
    "pagination": Object {
      "limit": 15,
      "next": 2,
      "page": 1,
      "pages": 4,
      "prev": null,
      "total": 56,
    },
  },
  "tags": Array [
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "count": Object {
        "posts": 7,
      },
      "description": null,
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "Getting Started",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "getting-started",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/getting-started/",
      "visibility": "public",
    },
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "count": Object {
        "posts": 2,
      },
      "description": "description",
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "bacon",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "bacon",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/bacon/",
      "visibility": "public",
    },
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "count": Object {
        "posts": 1,
      },
      "description": "description",
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "chorizo",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "chorizo",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/chorizo/",
      "visibility": "public",
    },
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "count": Object {
        "posts": 2,
      },
      "description": "description",
      "feature_image": "https://example.com/super_photo.jpg",
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "kitchen sink",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "kitchen-sink",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/kitchen-sink/",
      "visibility": "public",
    },
  ],
}
`;

exports[`Tags Content API Can include post count 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "1993",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Tags Content API Can limit tags to receive 1: [body] 1`] = `
Object {
  "meta": Object {
    "pagination": Object {
      "limit": 3,
      "next": 2,
      "page": 1,
      "pages": 19,
      "prev": null,
      "total": 56,
    },
  },
  "tags": Array [
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "description": null,
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "Getting Started",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "getting-started",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/getting-started/",
      "visibility": "public",
    },
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "description": "description",
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "bacon",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "bacon",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/bacon/",
      "visibility": "public",
    },
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "description": "description",
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "chorizo",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "chorizo",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/chorizo/",
      "visibility": "public",
    },
  ],
}
`;

exports[`Tags Content API Can limit tags to receive 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "1425",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Tags Content API Can request tags 1: [body] 1`] = `
Object {
  "meta": Object {
    "pagination": Object {
      "limit": 15,
      "next": 2,
      "page": 1,
      "pages": 4,
      "prev": null,
      "total": 56,
    },
  },
  "tags": Array [
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "description": null,
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "Getting Started",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "getting-started",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/getting-started/",
      "visibility": "public",
    },
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "description": "description",
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "bacon",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "bacon",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/bacon/",
      "visibility": "public",
    },
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "description": "description",
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "chorizo",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "chorizo",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/chorizo/",
      "visibility": "public",
    },
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "description": "description",
      "feature_image": "https://example.com/super_photo.jpg",
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "kitchen sink",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "kitchen-sink",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/kitchen-sink/",
      "visibility": "public",
    },
  ],
}
`;

exports[`Tags Content API Can request tags 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "1913",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Tags Content API Can request tags with limit=all 1: [body] 1`] = `
Object {
  "meta": Object {
    "pagination": Object {
      "limit": "all",
      "next": null,
      "page": 1,
      "pages": 1,
      "prev": null,
      "total": 4,
    },
  },
  "tags": Array [
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "description": null,
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "Getting Started",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "getting-started",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/getting-started/",
      "visibility": "public",
    },
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "description": "description",
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "bacon",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "bacon",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/bacon/",
      "visibility": "public",
    },
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "description": "description",
      "feature_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "chorizo",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "chorizo",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/chorizo/",
      "visibility": "public",
    },
    Object {
      "accent_color": null,
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "description": "description",
      "feature_image": "https://example.com/super_photo.jpg",
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "name": "kitchen sink",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "slug": "kitchen-sink",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "url": "http://127.0.0.1:2369/tag/kitchen-sink/",
      "visibility": "public",
    },
  ],
}
`;

exports[`Tags Content API Can request tags with limit=all 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "1918",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Tags Content API Can use multiple fields and have valid url fields 1: [body] 1`] = `
Object {
  "meta": Object {
    "pagination": Object {
      "limit": 15,
      "next": 2,
      "page": 1,
      "pages": 4,
      "prev": null,
      "total": 56,
    },
  },
  "tags": Array [
    Object {
      "name": "Getting Started",
      "url": "http://127.0.0.1:2369/tag/getting-started/",
    },
    Object {
      "name": "bacon",
      "url": "http://127.0.0.1:2369/tag/bacon/",
    },
    Object {
      "name": "chorizo",
      "url": "http://127.0.0.1:2369/tag/chorizo/",
    },
    Object {
      "name": "kitchen sink",
      "url": "http://127.0.0.1:2369/tag/kitchen-sink/",
    },
  ],
}
`;

exports[`Tags Content API Can use multiple fields and have valid url fields 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "366",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Tags Content API Can use single url field and have valid url fields 1: [body] 1`] = `
Object {
  "meta": Object {
    "pagination": Object {
      "limit": 15,
      "next": 2,
      "page": 1,
      "pages": 4,
      "prev": null,
      "total": 56,
    },
  },
  "tags": Array [
    Object {
      "url": "http://127.0.0.1:2369/tag/getting-started/",
    },
    Object {
      "url": "http://127.0.0.1:2369/tag/bacon/",
    },
    Object {
      "url": "http://127.0.0.1:2369/tag/chorizo/",
    },
    Object {
      "url": "http://127.0.0.1:2369/tag/kitchen-sink/",
    },
  ],
}
`;

exports[`Tags Content API Can use single url field and have valid url fields 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "287",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;
